rfcs/006-package-management.md

```diff
@@ -0,0 +1,359 @@
+
+### Alternative: explicit regeneration
+
+Instead of automatically updating the lock-file when the manifest changes,
+we could update it only on explicit commands: `plate eval` would use an
+old lock-file if it exists, while `plate lock` would read the manifest
+and update the lock-file. This alternative is potentially more efficient
+(as `plate eval` wouldn't need to re-read the manifest, and the manifests
+of path-dependencies, on every invocation), but makes it easy to accidentally
+use out-of-date packages.
+
+## Version compatibility and resolution
+
+How do we handle a package that gets imported multiple times in the dependency
+tree?
+
+For path and git dependencies, there isn't much choice. Dependencies from the
+registry are the most interesting.  Some languages (e.g. python) insist that
+each package resolves to a single version across the whole dependency tree.
+Other languages allow multiple versions, keeping track of which package in the
+dependency tree needs to import which version of a package.
+
+I think we want to allow multiple versions of a package; the alternative can be
+fragile and annoying. But then we need to figure out how many different versions
+to allow. There's a trade-off: if we allow pulling in a different version
+every time a package gets imported, solving the dependency graph is easy.
+But it increases the chance of getting incompatibilities at runtime: we might
+accidentally get a value from `util@1.1` and try to pass it to an incompatible
+function defined in `util@1.2`. Pulling in too many different versions also
+increases the total number of packages in the dependency graph.
+
+The current prototype uses a strategy similar to cargo: it divides package
+versions into semver-delimited "bins" and allows resolution to choose at most
+one version from each bin. That is, we can have a `util@2.2` and a `util@1.2` in
+the same dependency tree, but not a `util@1.2` and a `util@1.1`.
+
+### Question: how to handle pre-1.0 minor versions?
+
+Officially, semver says that pre-1.0 versions are mutually incompatible. If we
+follow this, pre-1.0 versions would never get binned together. `cargo` modifies
+the semver rules, allowing `0.x.y` versions to be binned together if they share
+the same `x`. Should we do the same?
+
+## The registry
+
+How should we manage the global registry? There's a potential for incurring
+substantial maintenance costs here, so we should be careful.
+
+We will provide a git repo, hard-coded to live at `github.com/tweag/nickel-mine`,
+to serve as the registry.
+This repo will contain the "index", but not the actual package contents. It
+will contain one file per package, each of which contains a line per version.
+Each entry specifies the location of the package (currently required to be on
+github) and its git tree hash. This ensures that packages are immutable, but it
+doesn't stop them from disappearing: we don't keep a copy of the actual package
+contents.
```

Comment:
This representation strategy means that the manifest files (hence the packages' metadata) isn't kept in the registry. It means that querying the registry can be quite expensive since it requires many calls to many Git repositories. Is this reasonable? (genuine question, I don't really have an opinion on this)

Reply:
This is also related to the thing I left out, that the registry index has enough information for dependency resolution. We don't store the whole manifest in the registry, but enough that we can do version resolution just by querying the index (and manifests from git/path dependencies, of course).

Reply:
Fair enough. I guess the index ought to include whatever metadata we want to be easily queryable.